images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/modules/expansion/i2c.py

```diff
@@ -0,0 +1,76 @@
+"""
+[module]
+enabled=true
+[end]
+"""
+
+import os
+import re
+import logging
+
+import pyftdi.i2c
+import binascii
+
+import asyncio
+import time
+
+from x1plus.utils import module_loader, module_docstring_parser
+
+logger = logging.getLogger(__name__)
+
+_registered_drivers = {}
+
+class I2cDriver():
+    DEVICE_DRIVERS = {}
+    
+    def __init__(self, daemon, config, ftdi_path, port_name):
+        self.ftdi_path = ftdi_path
+
+        self.i2c = pyftdi.i2c.I2cController()
+        self.i2c.configure(ftdi_path, frequency=50000)
+        
+        self.daemon = daemon
+        self.devices = []
+
+        for driver in _registered_drivers:
+            _registered_drivers[driver](self)
+
+        # I2C config format is just a dict of addresses -> devices
+        for address, devices in config.items():
+            address = int(address, 0)
+            
+            for device, subconfig in devices.items():
+                try:
+                    # Only load device with driver if requested by I2C config
+                    if driver := self.DEVICE_DRIVERS.get(device, None):
+                        self.devices.append(driver(address = address, i2c_driver = self, config = subconfig))
+                    else:
+                        logger.error(f"failed to initialize driver for {device}@0x{address:2x} on {ftdi_path}. Compatible driver not found")
```

Comment:
I think this no longer demand-loads, right?  also should this just look up into `_registered_drivers`?

Reply:
It no longer demand-loads/imports but it demand-instantiates. All the classes are loaded, but it only instantiates the driver on demand from the i2c port config.